File: input/InputDevice.cpp
//================================================================================
// InputDevice.cpp
// Authors:  Andrew Dieffenbach
// Date:     November 13, 2010
// Description:
//
//================================================================================

// INCLUDES ======================================================================

// Add the SDL header files for appropriate OS build
#include <SDL/SDL.h>

#include "puzl/input/InputDevice.h"	// InputDevice class header file for general input


// DEFINES =======================================================================

// TYPES =========================================================================

// PROTOTYPES ====================================================================

// EXTERNALS =====================================================================

// GLOBALS =======================================================================

// FUNCTIONS =====================================================================

//--------------------------------------------------------------------------------
// Name: InputDevice()
// Description:
// Default constructor
//--------------------------------------------------------------------------------
InputDevice::InputDevice()
{
	input = NULL;
	stateChange = NULL;
	stateChangeBufferSize = 0;
	numberOfStateChanges = 0;
	
	lastInputId = -1;
	lastInputType = 0;
}


//--------------------------------------------------------------------------------
// Name: ~InputDevice()
// Description:
// Default destructor
//--------------------------------------------------------------------------------
InputDevice::~InputDevice()
{

}


//--------------------------------------------------------------------------------
// Name: initialize()
// Description:
// 
//--------------------------------------------------------------------------------
int InputDevice::initialize()
{
	return 0;
}


//--------------------------------------------------------------------------------
// Name: shutdown()
// Description:
// 
//--------------------------------------------------------------------------------
int InputDevice::shutdown()
{
	return 0;
}


//--------------------------------------------------------------------------------
// Name: check()
// Description:
// 
//--------------------------------------------------------------------------------
bool InputDevice::check( int inputId )
{
	// NOTE: This implementation is prone to fail if state defines change
	// as it counts both INPUT_STATE_DOWN and INPUT_STATE_PRESSED as sharing
	// a common bit.
	return input[inputId]->state & INPUT_STATE_DOWN ? true : false;
}


//--------------------------------------------------------------------------------
// Name: getState()
// Description:
// 
//--------------------------------------------------------------------------------
int InputDevice::getState( int inputId )
{
	return input[inputId]->state;
}


//--------------------------------------------------------------------------------
// Name: getLastInputId()
// Description:
// 
//--------------------------------------------------------------------------------
int InputDevice::getLastInputId()
{
	return lastInputId;
}


//--------------------------------------------------------------------------------
// Name: update()
// Description:
// 
//--------------------------------------------------------------------------------
void InputDevice::update( SDL_Event *event )
{
	
}

//--------------------------------------------------------------------------------
// Name: age()
// Description:
// 
//--------------------------------------------------------------------------------
void InputDevice::age()
{
	int index;
	for( index = 0; index < numberOfStateChanges; index++ )
	{
		Input* tempStateChange = stateChange[index];
		
		if( tempStateChange->state == INPUT_STATE_RELEASED )
		{
			tempStateChange->state = INPUT_STATE_UP;
		}
		else
		if( tempStateChange->state == INPUT_STATE_PRESSED )
		{
			tempStateChange->state = INPUT_STATE_DOWN;
		}
	}
	
	numberOfStateChanges = 0;
	lastInputId = -1;
}

File: trunk/language/js/platform/html5/puzl/input/InputDevice.js
var INPUT_STATE_UP       = 0; // 00b
var INPUT_STATE_RELEASED = 2; // 10b
var INPUT_STATE_DOWN     = 1; // 01b
var INPUT_STATE_PRESSED  = 3; // 11b

/** @constructor */
function Input()
{
  this.state = 0;
  this.type  = 0;
  this.id    = 0;
  
  return this;
}

/** @constructor */
function InputDevice()
{
  /*this.input;
  this.stateChange;
  
  this.stateChangeBufferSize;
  this.numberOfStateChanges;
  
  this.lastInputId;
  this.lastInputType;*/

  // Constructor.
  this.stateChangeBufferSize = 0;
  this.numberOfStateChanges  = 0;

  this.lastInputId   = -1;
  this.lastInputType = 0;
}

InputDevice.prototype.initialize = function()
{
  return 0;
};

InputDevice.prototype.shutdown = function()
{
  return 0;
};

InputDevice.prototype.check = function( inputId )
{
  return ( this.input[inputId].state & INPUT_STATE_DOWN ) === 0 ? false : true;
};

InputDevice.prototype.getState = function( inputId )
{
  return this.input[inputId].state;
};

InputDevice.prototype.setState = function( inputId, state )
{
  var thisInput = this.input[inputId];
  if( state === INPUT_STATE_PRESSED )
  {
    var currentState = thisInput.state;
    if( ( currentState & INPUT_STATE_DOWN ) > 0 )
    {
      return;
    }

    this.lastInputId = inputId;
  }
  
  thisInput.state = state;

  var stateChangeIndex = this.numberOfStateChanges++;
  
  var thisStateChange = this.stateChange;
  if( thisStateChange.length < stateChangeIndex + 1 )
  {
    thisStateChange.push( thisInput );
  }
  else
  {
    thisStateChange[stateChangeIndex] = thisInput;
  }
};

InputDevice.prototype.getLastInputId = function()
{
  return this.lastInputId;
};

InputDevice.prototype.age = function()
{
  var numberOfStateChanges = this.numberOfStateChanges;
  if( numberOfStateChanges > 0 )
  {
    var thisStateChange = this.stateChange;
    
    var tempState;
    var tempStateChange;
    var index;
    for( index = 0; index < numberOfStateChanges; index++ )
    {
      tempStateChange = thisStateChange[index];
      tempState = tempStateChange.state;
      if( tempState === INPUT_STATE_RELEASED )
      {
        tempStateChange.state = INPUT_STATE_UP;
      }
      else
      if( tempState === INPUT_STATE_PRESSED )
      {
        tempStateChange.state = INPUT_STATE_DOWN;
      }
    }

    this.numberOfStateChanges = 0;
    this.lastInputId          = -1;
  }
};
